README.md
# password_validation_plus

A Flutter plugin that offers a password validation widget to ensure strong password requirements in your Flutter applications. The widget enforces 5 crucial rules for password strength, accompanied by a visual progress indicator to guide users in creating secure passwords.

![password_validation_plus_screenshot_1](https://github.com/nishalsehan/password_validation_plus/assets/44578204/3076bde8-94af-4afa-a22b-74e983db23b0)

## Features

- **Password Strength Rules**: The widget enforces 5 crucial rules for strong passwords:
  1. Minimum Length
  2. Uppercase Letters
  3. Lowercase Letters
  4. Special Characters
  5. Numbers (Digits)
  
- **Real-time Feedback**: Provides real-time feedback to users as they type their password, indicating their progress in meeting the specified requirements.

- **Password Strength Indicator**: Visual progress indicator that visually represents the strength of the entered password.

- **Customization**: Easily customize the password validation rules and the appearance of the widget to match your application's requirements.

![Password Validation Widget Demo](https://github.com/nishalsehan/password_validation_plus/assets/44578204/69cc3675-98c5-46fd-9517-d6fd8e1f4827)

## Installation

Add the following dependency to your `pubspec.yaml` file:

```yaml
dependencies:
  password_validation_plus: ^0.0.1
```

Run `flutter pub` get to install the package.

</br>

## Usage

Import the package in your Dart code:

```dart
import 'package:password_validation_plus/password_validation_plus.dart';
```
<br>
Use the PasswordValidationPlus widget in your widget tree:

```dart
PasswordValidationPlus(
  textController: controller,//TextEditingController
)
```
</br>

## Customization

You can customize the password validation widget by providing different parameters to the `PasswordValidationPlus` constructor:

```dart
PasswordValidationPlus(
  textController: controller,
  maxLength: 12,
  minLength: 8,
  successIcon: Iconsax.tick_circle,
  unSuccessIcon: Iconsax.close_circle,
  successWidget: SvgPicture.asset(
    "assets/icons/tick.svg",
    height: 24,
  ),
  unSuccessWidget: SvgPicture.asset(
    "assets/icons/close.svg",
    height: 24,
  ),
  textSize: 16,
  hideRules: true,
),
```

- `textController`: Set the TextEditingController of your password text field here.
- `minLength`: Set the minimum required length for the password.
- `maxLength`: Set the maximum allowed length for the password.
- `hideRules`: Set to true to hide the password validation rules.
- `successIcon`: Customize the icon displayed for a successful validation rule.
- `unSuccessIcon`: Customize the icon displayed for an unsuccessful validation rule.
- `successWidget`: Customize the widget displayed for a successful validation rule.
- `unSuccessWidget`: Customize the widget displayed for an unsuccessful validation rule.
- `textSize`: Adjust the text size of the validation rule and icon.
<br>

## Example

For a complete example of how to use this plugin, check out the [example](https://github.com/nishalsehan/password_validation_plus/tree/main/example) folder in this repository.
</br>

## Issues and Feedback

If you encounter any issues or have suggestions for improvement, please open an issue on our [GitHub repository](https://github.com/nishalsehan/password_validation_plus/issues). We welcome your feedback!
</br>

## License

This project is licensed under the MIT License - see the [LICENSE](https://github.com/nishalsehan/password_validation_plus/blob/main/LICENSE) file for details.


CHANGELOG.md
## 0.0.1

* TODO: PasswordValidation Plugin released

## 0.0.2

* TODO: dartdoc created

## 0.0.3

* TODO: sdk version changed to support lower sdks

test/password_validation_plus_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:password_validation_plus/password_validation_plus.dart';

void main() {

}


lib/password_validation_plus.dart
library password_validation_plus;

import 'package:flutter/material.dart';

class PasswordValidationPlus extends StatefulWidget {
  final TextEditingController textController;
  final int? maxLength;
  final int minLength;
  final IconData? successIcon;
  final IconData? unSuccessIcon;
  final Widget? successWidget;
  final Widget? unSuccessWidget;
  final double textSize;
  final bool hideRules;
  const PasswordValidationPlus(
      {super.key,
      required this.textController,
      this.maxLength,
      this.minLength = 8,
      this.successIcon,
      this.unSuccessIcon,
      this.successWidget,
      this.unSuccessWidget,
      this.textSize = 14,
      this.hideRules = false});

  @override
  State<PasswordValidationPlus> createState() => PasswordValidationPlusState();
}

class PasswordValidationPlusState extends State<PasswordValidationPlus> {
  RegExp numReg = RegExp(r".*[0-9].*"); //use to find password text contains at least one digit
  RegExp simpleReg = RegExp(r".*[a-z].*"); //use to find password text contains at least one simple letter
  RegExp capitalReg = RegExp(r".*[A-Z].*"); //use to find password text contains at least one CAPITAL letter
  RegExp symbolsReg = RegExp(r'.*[!@#\$&*~].*'); //use to find password text contains at least one symbol (!, @, #, \, $, &, *, ~)
  double _strength = 0;

  @override
  void initState() {
    super.initState();
    //add a listener to listen on password field changes
    widget.textController.addListener(calculateStrength);
  }

  @override
  Widget build(BuildContext context) {
    Size size = MediaQuery.of(context).size;
    return SizedBox(
      width: size.width * 0.85,
      child: Column(
        children: [
          SizedBox(height: size.height * 0.005),
          //show the progress of the password strength
          LinearProgressIndicator(
            value: _strength,
            backgroundColor: Colors.grey[300],
            color: _strength <= 0.2
                ? Colors.red
                : _strength == 0.4
                    ? Colors.yellow
                    : _strength == 0.6000000000000001
                        ? Colors.amber
                        : _strength == 0.8
                            ? Colors.orange
                            : Colors.green,
            minHeight: 5,
          ),
          if (!widget.hideRules)
            Column(
              children: [
                SizedBox(height: size.height * 0.015),
                conditionText(
                    checkLength(),
                    widget.maxLength != null
                        ? 'Contain between ${widget.minLength}-${widget.maxLength} Characters'
                        : 'Contain at least ${widget.minLength} Characters'),
                SizedBox(height: size.height * 0.005),
                conditionText(capitalReg.hasMatch(widget.textController.text),
                    "At least one CAPITAL letter"),
                SizedBox(height: size.height * 0.005),
                conditionText(simpleReg.hasMatch(widget.textController.text),
                    "At least one simple letter"),
                SizedBox(height: size.height * 0.005),
                conditionText(numReg.hasMatch(widget.textController.text),
                    "At least one number"),
                SizedBox(height: size.height * 0.005),
                conditionText(symbolsReg.hasMatch(widget.textController.text),
                    "At least one special character"),
              ],
            )
        ],
      ),
    );
  }

  //calculate the password strength base on 5 conditions
  calculateStrength() {
    double total = 0;
    if (checkLength()) {
      total += 0.2;
    }
    if (capitalReg.hasMatch(widget.textController.text)) {
      total += 0.2;
    }
    if (simpleReg.hasMatch(widget.textController.text)) {
      total += 0.2;
    }
    if (numReg.hasMatch(widget.textController.text)) {
      total += 0.2;
    }
    if (symbolsReg.hasMatch(widget.textController.text)) {
      total += 0.2;
    }
    setState(() {
      _strength = total;
    });
  }

  Widget conditionText(bool condition, String text) {
    return Row(
      crossAxisAlignment: CrossAxisAlignment.center,
      children: [
        Container(
          alignment: Alignment.centerLeft,
          child: condition
              ? widget.successWidget ??
                  Icon(
                    widget.successIcon ?? Icons.check,
                    color: Colors.green,
                    size: 14,
                  )
              : widget.unSuccessWidget ??
                  Icon(
                    widget.unSuccessIcon ?? Icons.clear,
                    color: Colors.red,
                    size: 14,
                  ),
        ),
        const SizedBox(width: 12),
        SizedBox(
          child: Text(
            text,
            textAlign: TextAlign.left,
            style: TextStyle(
              fontSize: 14,
              color: condition ? Colors.green : Colors.red,
              height: 1.5,
            ),
          ),
        )
      ],
    );
  }

  checkLength() {
    if (widget.maxLength != null) {
      return widget.textController.text.length >= widget.minLength &&
          widget.textController.text.length <= widget.maxLength!; //check maximum character length if maxLength field is not null
    } else {
      return widget.textController.text.length >= widget.minLength;
    }
  }
}


example/lib/widgets/password_text_field.dart
import 'package:flutter/material.dart';
import 'package:iconsax/iconsax.dart';


class PasswordTextField extends StatefulWidget {
  final String hint;
  final TextEditingController controller;
  final void Function(void Function()) setState;

  const PasswordTextField(
      {super.key,
      required this.hint,
      required this.controller,
      required this.setState});

  @override
  State<StatefulWidget> createState() => PasswordTextFieldState();
}

class PasswordTextFieldState extends State<PasswordTextField> {
  late FocusNode focusNode = FocusNode();
  bool _passwordVisible = false;

  @override
  Widget build(BuildContext context) {
    Size size = MediaQuery.of(context).size;
    return Container(
      width: size.width * 0.9,
      alignment: Alignment.bottomCenter,
      child: TextFormField(
        obscureText: !_passwordVisible,
        keyboardType: TextInputType.text,
        controller: widget.controller,
        focusNode: focusNode,
        onChanged: (value) {
          widget.setState(() {});
        },
        showCursor: false,
        decoration: InputDecoration(
          isDense: true,
          hintText: widget.hint,
          enabledBorder: OutlineInputBorder(
            borderSide: const BorderSide(color: Colors.black26, width: 1),
            borderRadius: BorderRadius.circular(12),
          ),
          border: OutlineInputBorder(
            borderSide: const BorderSide(color: Colors.black26, width: 1),
            borderRadius: BorderRadius.circular(12),
          ),
          focusedBorder: OutlineInputBorder(
            borderSide: const BorderSide(color: Colors.black38, width: 1),
            borderRadius: BorderRadius.circular(12),
          ),
          suffixIcon: InkWell(
            splashColor: Colors.transparent,
            highlightColor: Colors.transparent,
            child: Container(
              padding: EdgeInsets.only(
                  left: size.width * 0.02, right: size.width * 0.01),
              child: Icon(_passwordVisible ? Iconsax.eye3 : Iconsax.eye_slash,
                  color: focusNode.hasFocus ? Colors.black54 : Colors.black26,
                  size: size.height * 0.022),
            ),
            onTap: () {
              setState(() {
                _passwordVisible = !_passwordVisible;
              });
            },
          ),
          suffixIconConstraints: BoxConstraints(
            minWidth: size.width * 0.15,
          ),
          hintStyle: Theme.of(context).textTheme.labelLarge!.copyWith(
              color: Colors.black38,
              fontWeight: FontWeight.w500,
              height: 1.4,
              letterSpacing: 1.5),
          counterText: "",
          contentPadding: EdgeInsets.symmetric(
              vertical: size.height * 0.018, horizontal: size.width * 0.04),
        ),
        cursorColor: Colors.black26,
        style: Theme.of(context).textTheme.labelLarge!.copyWith(
            color: Colors.black87,
            fontWeight: FontWeight.w500,
            height: 1.4,
            letterSpacing: 1.5),
      ),
    );
  }

  @override
  void initState() {
    focusNode.addListener(() {
      setState(() {});
    });
    super.initState();
  }
}


example/lib/screens/home_page.dart
import 'package:flutter/material.dart';
import 'package:iconsax/iconsax.dart';
import 'package:password_validation_plus/password_validation_plus.dart';

import '../widgets/widget_barrel.dart';

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});
  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  TextEditingController controller = TextEditingController();
  @override
  Widget build(BuildContext context) {
    final size = MediaQuery.of(context).size;
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Column(
        children: [
          SizedBox(
            width: size.width,
            height: size.height * 0.01,
          ),
          SizedBox(
            width: size.width * 0.9,
            child: Text(
              'Set a Password',
              style: Theme.of(context)
                  .textTheme
                  .titleLarge!
                  .copyWith(fontWeight: FontWeight.w600),
            ),
          ),
          SizedBox(
            height: size.height * 0.02,
          ),
          SizedBox(
            width: size.width * 0.9,
            child: Text(
              'Please create a secure password including the following criteria below',
              style: Theme.of(context).textTheme.titleMedium,
            ),
          ),
          SizedBox(
            height: size.height * 0.02,
          ),
          PasswordTextField(
            hint: 'Password',
            controller: controller,
            setState: setState,
          ),
          SizedBox(
            height: size.height * 0.005,
          ),
          PasswordValidationPlus(
            textController: controller,
            maxLength: 12,
            minLength: 8,
            successIcon: Iconsax.tick_circle,
            unSuccessIcon: Iconsax.close_circle,
            textSize: 16,
          ),
        ],
      ),
    );
  }
}
